[FLINK-31409][hive] Hive dialect should use public interfaces

luoyuxia opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
This is to make Hive dialect use public interfaces provided by pluggable dialect.
After this, we can then remove provided dependency of flink-table-planner in FLINK-31413


## Brief change log
The basic idea comes from [FLIP-216](https://cwiki.apache.org/confluence/display/FLINK/FLIP-216%3A++Introduce+pluggable+dialect+and+plan+for+migrating+Hive+dialect).
-  In [first commit](https://github.com/apache/flink/pull/22166/commits/b289fd0abdd0b7fcadb34369416bcde0f7d6efb8),  we introduce a slim module named `flink-table-calcite-bridge` which ships calcite dependency and provide a `CalciteContext` usded by other plugin module. In the next pr, we can then make hive-connector depend on `flink-table-calcite-bridge`
- In [second commit](https://github.com/apache/flink/pull/22166/commits/c6c4883805d75a8ddaab4dd8f33e335add5d9758), we introudce or expose some public interfaces like `DialectFactory`/`CatalogRegistry` for pulgin dialect .
Something to add, in here we remove `ExtendedOperationExecutor`, as a result of which we wrap the logic in `HiveExtendedOperationExecutor` to `HiveExecutableOperation`.   `HiveExecutableOperation` will delegate the execution logic to `HiveExtendedOperationExecutor` in method `execute`.  It's some hack logic and will be refactor in other pr.

- In [thrid commit](https://github.com/apache/flink/pull/22166/commits/7612ccaed9ca85b534014ca4dba9d442fdf5c278), we make HiveParser use public interface `CatalogRegistry` and an internal  inteface `CalciteContext` instead of internal class/interfaces like `CatalogManager`/`FlinkPlannerImpl`/`PlannerContext`, ...


## Verifying this change
The change has been verified by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

snuyanzin commented on flink-table/flink-table-calcite-bridge/pom.xml:
```diff
@@ -0,0 +1,145 @@
+			<version>${calcite.version}</version>
+			<exclusions>
+				<!--
+				"mvn dependency:tree" as of Calcite 1.26.0:
+				[INFO] +- org.apache.calcite:calcite-core:jar:1.26.0:compile
+				[INFO] |  +- org.apache.calcite:calcite-linq4j:jar:1.26.0:compile
+				[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.12.1:compile
+				[INFO] |  +- org.apiguardian:apiguardian-api:jar:1.1.0:compile
+				[INFO] |  +- com.esri.geometry:esri-geometry-api:jar:2.2.0:runtime
+				[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.12.1:runtime
+				[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.12.1:runtime
+				[INFO] |  +- com.jayway.jsonpath:json-path:jar:2.4.0:runtime
+				[INFO] |  |  \- net.minidev:json-smart:jar:2.3:runtime
+				[INFO] |  |     \- net.minidev:accessors-smart:jar:1.2:runtime
+				[INFO] |  |        \- org.ow2.asm:asm:jar:5.0.4:runtime
+				[INFO] |  +- commons-codec:commons-codec:jar:1.10:runtime
+				[INFO] |  +- org.apache.commons:commons-lang3:jar:3.3.2:compile
+				[INFO] |  \- commons-io:commons-io:jar:2.4:compile
```
it should be 1.29.0 dependencies like at https://github.com/apache/flink/pull/22166/files#diff-6d81d722fce4e303d49b2136ae8e137f74ed3f4ece4d2a990445bcf6a2388392L132-L147
```suggestion
								"mvn dependency:tree" as of Calcite 1.29.0:
				[INFO] +- org.apache.calcite:calcite-core:jar:1.29.0:compile
				[INFO] |  +- org.apache.calcite:calcite-linq4j:jar:1.29.0:compile
				[INFO] |  +- com.esri.geometry:esri-geometry-api:jar:2.2.0:compile
				[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.13.4:compile
				[INFO] |  +- org.apache.calcite.avatica:avatica-core:jar:1.20.0:compile
				[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.13.4:compile
				[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.13.4.2:compile
				[INFO] |  +- com.jayway.jsonpath:json-path:jar:2.4.0:runtime
				[INFO] |  |  \- net.minidev:json-smart:jar:2.3:runtime
				[INFO] |  |     \- net.minidev:accessors-smart:jar:1.2:runtime
				[INFO] |  |        \- org.ow2.asm:asm:jar:5.0.4:runtime
				[INFO] |  +- commons-codec:commons-codec:jar:1.15:runtime
				[INFO] |  \- commons-io:commons-io:jar:2.11.0:compile
```

luoyuxia commented:
@wuchong @lincoln-lil Could you please help review when you are free.

lincoln-lil commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java:
```diff
@@ -134,9 +138,22 @@ private ResultFetcher(
+            resultRows = tableResultInternal.collectInternal();
+            rowDataToStringConverter = tableResultInternal.getRowDataToStringConverter();
+        } else {
+            // sometime, the tableResult maybe not an instance of TableResultInternal
```
nit: 'sometimes'

lincoln-lil reviewed (commented):
@luoyuxia thanks for driving the hive decoupling! Overall looks good to me, and I've left some minor comments.

luoyuxia commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:
```diff
@@ -169,17 +171,46 @@ abstract class PlannerBase(
+        Optional.of(ApiExpressionUtils.tableRef(path, queryOperation))
+      }
+    } catch {
+      case _: SqlParserException => Optional.empty()
```
It's copied from `TableEnvironmentImpl`, I think we can also copy comment to here.
Haven't find any test cases related to this exception path. Considering it's copied from existing codebase , may be we can add test in a dedicated pr if need.

lincoln-lil replied:
sounds good to create another followup issue to add a related test

lincoln-lil reviewed (approved):
LGTM +1

lincoln-lil commented:
@luoyuxia seems the latest change failed to pass the style check, PTAL
otherwise +1

luoyuxia commented:
@flinkbot run azure

wuchong commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/DialectFactory.java:
```diff
@@ -36,52 +34,22 @@
     Parser create(Context context);
 
+    /** Create an extended operation executor. */
     default ExtendedOperationExecutor createExtendedOperationExecutor(Context context) {
```
I'm thinking of removing the `ExtendedOperationExecutor` because we have introduced `ExecutableOperation` which is a self-contained way to support the operation execution. In this way, we can rename `DialectFactory` back to `ParserFactory` to align with the original FLIP proposal.

luoyuxia replied:
Removing `ExtendedOperationExecutor` is good idea. It makes everything simple. I have updated it for removing and renming.
Hope the internal `ExecutableOperation` interface won't change frequently

wuchong commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java:
```diff
@@ -134,9 +138,22 @@ private ResultFetcher(
+        } else {
+            // sometimes, the tableResult maybe not an instance of TableResultInternal
+            // in the case of pluggable dialect implements method
+            // ExtendedOperationExecutor#executeOperation, and it doesn't return TableResultInternal
```
Which case doesn't return `TableResultInternal`?

luoyuxia replied:
Previous, we have `ExtendedOperationExecutor#executeOperation` marked as `PublicEvolving`, so the result of this method should be `TableResult` which is also marked as `PublicEvolving`.  Since this interface don't require implementer to return `TableResultInternal`, the implementer can return anything that implements `TableResult`.

But now, I have removed `ExtendedOperationExecutor` and reverted this change.

wuchong commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentInternal.java:
```diff
@@ -74,15 +75,15 @@ public interface TableEnvironmentInternal extends TableEnvironment {
      * @return the content of the execution result.
      */
-    TableResultInternal executeInternal(Operation operation);
+    TableResult executeInternal(Operation operation);
```
Why do we need to change this?

luoyuxia replied:
Previous, we have `ExtendedOperationExecutor#executeOperation` marked as `PublicEvolving`, so the result of this method should be TableResult which is also marked as `PublicEvolving`. So we should also change it to `TableResult` in here.

Now, I revert this changes since we remove `ExtendedOperationExecutor`.

wuchong commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:
```diff
@@ -169,17 +171,53 @@ abstract class PlannerBase(
     parser
   }
 
+  private def getOperationTreeBuilder: OperationTreeBuilder = {
+    OperationTreeBuilderImpl.create(
+      tableConfig,
+      classLoader,
+      functionCatalog.asLookup(f => getParser.parseIdentifier(f)),
+      catalogManager.getDataTypeFactory,
+      (path: String) => getTableReferenceExpression(path),
+      (s: String, inputRowType: RowType, outputType) =>
+        getParser.parseSqlExpression(s, inputRowType, outputType),
+      isStreamingMode
+    )
+  }
```
The creation of OperationTreeBuilderImpl is super complex, and the logic is duplicated. This is difficult to maintain logic consistency in the future. 

Besides, the extracted `OperationTreeBuilder` seems incomplete (e.g, join) and is never used. I would prefer to introduce the `OperationTreeBuilder` when we indeed need this, otherwise, it only increases the overhead of maintenance. What do you think?

luoyuxia replied:
I do agree. We can add it in the future when we do need it. That does make everything simple.

Btw, the reason that `OperationTreeBuilder` miss join is the `JoinType` is not marked as `PublicEvolving `, so we can't expose it in `OperationTreeBuilder`.

luoyuxia commented:
I reoriginize the commits intead of appending for this change make append commit hard to track. I also update the `brief change log` in the description for this pr.

wuchong reviewed (commented):
Thank you for the update. I think this is in good shape. During reviewing the code, I find the methods of `CatalogRegistry` are not smooth to use. For example, 
1. the `ResolvedCatalogBaseTable` vs `ContextResolvedTable` mentioned above, 
2. the `Optional` return type is trivial, we always throw exceptions when working with an empty Optional result. 
3. it's a long method chain to get the current catalog, but this is a frequent call. 

I'm thinking about revisiting the design of `CatalogRegistry` into this:

```java
/** A catalog registry for dealing with catalogs. */
@PublicEvolving
public interface CatalogRegistry {
    /** Get the current database. */
    String getCurrentDatabase();

    /** Gets the current catalog. */
    String getCurrentCatalog();

    /**
     * Returns the full name of the given table path, this name may be padded with current
     * catalog/database name based on the {@code identifier's} length.
     *
     * @param identifier an unresolved identifier
     * @return a fully qualified object identifier
     */
    ObjectIdentifier qualifyIdentifier(UnresolvedIdentifier identifier);

    /** Gets the current catalog. */
    Catalog getCurrentCatalog();

    /**
     * Gets a catalog by name.
     *
     * @param catalogName name of the catalog to retrieve
     * @return the requested catalog
     * @throws CatalogNotExistException if the catalog does not exist
     */
    Catalog getCatalogOrError(String catalogName);

    /**
     * Gets a fully qualified and resolved table with context by name. If the path is not yet fully
     * qualified use {@link #qualifyIdentifier(UnresolvedIdentifier)} first.
     *
     * @param objectIdentifier full path of the table to retrieve
     * @return resolved table with context that the path points to
     * @throws TableNotExistException if the table does not exist
     */
    ContextResolvedTable getTableOrError(ObjectIdentifier objectIdentifier);

    /**
     * Return whether the table with a fully qualified table path is temporary or not.
     *
     * @param objectIdentifier full path of the table
     * @return the table is temporary or not.
     */
    boolean isTemporaryTable(ObjectIdentifier objectIdentifier);

    /**
     * Retrieves a partition with a fully qualified table path and partition spec.
     *
     * @param tableIdentifier full path of the table to retrieve
     * @param partitionSpec full partition spec
     * @return partition in the table.
     */
    Optional<CatalogPartition> getPartition(
            ObjectIdentifier tableIdentifier, CatalogPartitionSpec partitionSpec);
    
    /**
     * Retrieves a partition with a fully qualified table path and partition spec.
     *
     * @param tableIdentifier full path of the table to retrieve
     * @param partitionSpec full partition spec
     * @return partition in the table.
     * @throws PartitionNotExistException if the partition does not exist
     */
    CatalogPartition getPartitionOrError(
            ObjectIdentifier tableIdentifier, CatalogPartitionSpec partitionSpec);
}
```

What do you think?

luoyuxia commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserDMLHelper.java:
```diff
@@ -331,20 +333,48 @@ public Operation createInsertOperation(HiveParserCalcitePlanner analyzer, RelNod
                 Collections.emptyMap());
     }
 
+    private ContextResolvedTable getContextResolvedTable(ObjectIdentifier tableIdentifier) {
```
Yes, we still needs to depend on `ContextResolvedTable` for it's need to construct `SinkModifyOperation` which is also `@Internal`. In fact, Hive dialect still use many internal operations. Maybe we need to expose them to public then,  but it may be another topic. 
Considering our plan is decoupling Hive with flink-table-planner in the first, I think we can keep them to see how much these internal  interfaces/classes will affect the maintainance/development of Hive dialect to take further steps.
Btw, it make me be aware of the title of the pr/jira is not correct,  it should be `Hive dialect shouldn't use interfaces in table-planner to build LogicPlan`.

luoyuxia commented:
@wuchong Thanks for your review. I have addressed your comments in [9e508a](https://github.com/apache/flink/pull/22166/commits/9e508ab847053ae91f65afa5bca929a079a76e79),
[9a5350](https://github.com/apache/flink/pull/22166/commits/9a53507611754f1e95be223eac19368430aa4818).

wuchong reviewed (approved):
@luoyuxia thank you for the update. LGTM.

luoyuxia commented:
Thanks all for the review. :-)
